src/python-common/ceph/deployment/service_spec.py

```diff
@@ -1067,7 +1067,7 @@ def to_json(self):
                 continue
             if hasattr(val, 'to_json'):
                 val = val.to_json()
-            if val:
+            if val or val == False:
```

Comment:
Perhaps, `if val is not None` would be a better choice here? Or alternatively, `if val or isinstance(val, bool)`?  Either way, it's not a recommended practice to use `==` for booleans - `is` would be preferred.

I highly recommend writing at least one test - either unit test or functional (teuthology) because it will help prevent future regressions and it aids us in reviewing the fix and understanding what exactly is getting fixed.

Reply:
this also broke a bunch of tests
```
FAILED cephadm/tests/test_cephadm.py::TestCephadm::test_service_ls - Assertio...
FAILED cephadm/tests/test_cephadm.py::TestCephadm::test_service_ls_service_type_flag
FAILED cephadm/tests/test_migration.py::test_migrate_rgw_spec[rgw_spec_store_entry0-True]
FAILED cephadm/tests/test_spec.py::test_spec_octopus[spec_json0] - AssertionE...
FAILED cephadm/tests/test_spec.py::test_spec_octopus[spec_json1] - AssertionE...
FAILED cephadm/tests/test_spec.py::test_spec_octopus[spec_json2] - AssertionE...
FAILED cephadm/tests/test_spec.py::test_spec_octopus[spec_json3] - AssertionE...
FAILED cephadm/tests/test_spec.py::test_spec_octopus[spec_json4] - AssertionE...
FAILED cephadm/tests/test_spec.py::test_spec_octopus[spec_json5] - AssertionE...
FAILED cephadm/tests/test_spec.py::test_spec_octopus[spec_json6] - AssertionE...
FAILED cephadm/tests/test_spec.py::test_spec_octopus[spec_json7] - AssertionE...
FAILED cephadm/tests/test_spec.py::test_spec_octopus[spec_json8] - AssertionE...
FAILED orchestrator/tests/test_orchestrator.py::test_yaml - assert 'service_t...
FAILED orchestrator/tests/test_orchestrator.py::test_orch_ls - AssertionError...
```
```
FAILED ceph/tests/test_service_spec.py::test_yaml[service_type: crash\nservice_name: crash\nplacement:\n  host_pattern: '*'\n]
FAILED ceph/tests/test_service_spec.py::test_yaml[service_type: crash\nservice_name: crash\nplacement:\n  host_pattern: '*'\nunmanaged: true\n]
FAILED ceph/tests/test_service_spec.py::test_yaml[service_type: crash\nservice_name: crash\nplacement:\n  host_pattern:\n    pattern: Foo[0-9]|Bar[0-9]\n    pattern_type: regex\n]
FAILED ceph/tests/test_service_spec.py::test_yaml[service_type: rgw\nservice_id: default-rgw-realm.eu-central-1.1\nservice_name: rgw.default-rgw-realm.eu-central-1.1\nplacement:\n  hosts:\n  - ceph-001\nnetworks:\n- 10.0.0.0/8\n- 192.168.0.0/16\nspec:\n  rgw_frontend_type: civetweb\n  rgw_realm: default-rgw-realm\n  rgw_zone: eu-central-1\n]
FAILED ceph/tests/test_service_spec.py::test_yaml[service_type: osd\nservice_id: osd_spec_default\nservice_name: osd.osd_spec_default\nplacement:\n  host_pattern: '*'\nspec:\n  data_devices:\n    model: MC-55-44-XZ\n  db_devices:\n    model: SSD-123-foo\n  filter_logic: AND\n  objectstore: bluestore\n  wal_devices:\n    model: NVME-QQQQ-987\n]
FAILED ceph/tests/test_service_spec.py::test_yaml[service_type: alertmanager\nservice_name: alertmanager\nspec:\n  port: 1234\n  user_data:\n    default_webhook_urls:\n    - foo\n]
FAILED ceph/tests/test_service_spec.py::test_yaml[service_type: grafana\nservice_name: grafana\nspec:\n  anonymous_access: true\n  port: 1234\n  protocol: https\n]
FAILED ceph/tests/test_service_spec.py::test_yaml[service_type: grafana\nservice_name: grafana\nspec:\n  anonymous_access: true\n  initial_admin_password: secure\n  port: 1234\n  protocol: https\n]
FAILED ceph/tests/test_service_spec.py::test_yaml[service_type: ingress\nservice_id: rgw.foo\nservice_name: ingress.rgw.foo\nplacement:\n  hosts:\n  - host1\n  - host2\n  - host3\nspec:\n  backend_service: rgw.foo\n  first_virtual_router_id: 50\n  frontend_port: 8080\n  monitor_port: 8081\n  virtual_ip: 192.168.20.1/24\n]
FAILED ceph/tests/test_service_spec.py::test_yaml[service_type: nfs\nservice_id: mynfs\nservice_name: nfs.mynfs\nspec:\n  idmap_conf:\n    general:\n      local-realms: domain.org\n    mapping:\n      nobody-group: nfsnobody\n      nobody-user: nfsnobody\n  port: 1234\n]
FAILED ceph/tests/test_service_spec.py::test_yaml[service_type: iscsi\nservice_id: iscsi\nservice_name: iscsi.iscsi\nnetworks:\n- ::0/8\nspec:\n  api_password: admin\n  api_port: 5000\n  api_user: admin\n  pool: pool\n  trusted_ip_list:\n  - ::1\n  - ::2\n]
FAILED ceph/tests/test_service_spec.py::test_yaml[service_type: container\nservice_id: hello-world\nservice_name: container.hello-world\nspec:\n  args:\n  - --foo\n  bind_mounts:\n  - - type=bind\n    - source=lib/modules\n    - destination=/lib/modules\n    - ro=true\n  dirs:\n  - foo\n  - bar\n  entrypoint: /usr/bin/bash\n  envs:\n  - FOO=0815\n  files:\n    bar.conf:\n    - foo\n    - bar\n    foo.conf: 'foo\n\n      bar'\n  gid: 2000\n  image: docker.io/library/hello-world:latest\n  ports:\n  - 8080\n  - 8443\n  uid: 1000\n  volume_mounts:\n    foo: /foo\n]
FAILED ceph/tests/test_service_spec.py::test_yaml[service_type: snmp-gateway\nservice_name: snmp-gateway\nplacement:\n  count: 1\nspec:\n  credentials:\n    snmp_community: public\n  snmp_destination: 192.168.1.42:162\n  snmp_version: V2c\n]
FAILED ceph/tests/test_service_spec.py::test_yaml[service_type: snmp-gateway\nservice_name: snmp-gateway\nplacement:\n  count: 1\nspec:\n  auth_protocol: MD5\n  credentials:\n    snmp_v3_auth_password: mypassword\n    snmp_v3_auth_username: myuser\n  engine_id: 8000C53F00000000\n  port: 9464\n  snmp_destination: 192.168.1.42:162\n  snmp_version: V3\n]
FAILED ceph/tests/test_service_spec.py::test_yaml[service_type: snmp-gateway\nservice_name: snmp-gateway\nplacement:\n  count: 1\nspec:\n  credentials:\n    snmp_v3_auth_password: mypassword\n    snmp_v3_auth_username: myuser\n    snmp_v3_priv_password: mysecret\n  engine_id: 8000C53F00000000\n  privacy_protocol: AES\n  snmp_destination: 192.168.1.42:162\n  snmp_version: V3\n]
FAILED ceph/tests/test_service_spec.py::test_yaml[service_type: grafana\nservice_name: grafana\nplacement:\n  count: 1\nspec:\n  anonymous_access: false\n  initial_admin_password: password\n  protocol: https\n]
FAILED ceph/tests/test_service_spec.py::test_repr - assert "ServiceSpec....nt...
```

so this could start by taking a look at those